### ISSUE TITLE
ClearDomainEvents/Outbox

### ISSUE THREAD

**Issue by batesm** [m-jovanovic/event-reminder#2]
Hi. I queried this on one of your tweets.

If you call ClearDomainEvents on SavingChanges with the Outbox pattern but an exception is thrown (changes fail to save at the first attempt) then isn't it true that the domain events are not restored, ready for a retry.

Surely, after a successful retry the domain events would be lost.

**Comment by m-jovanovic** [comment]
When you say retry, do you mean a retry of `SaveChanges` or `SaveChangesAsync`?

**Comment by batesm** [comment]
Both. I would assume dispatching of domain events to Outbox would occur regardless of whether the async or non-async version was used.

Unless I am missing something, clearing domain events before committing the transaction could be a problem if you retry SaveChanges (Async) after failure on the same instance of DbContext. 

(By the way, apologies that the particular question I am asking is not entirely relevant to code in this repo. The code here publishes directly to mediator as opposed to saving outbox messages to the database)

**Comment by m-jovanovic** [comment]
I assume your initial thesis is correct. An automatic retry would cause the domain events to be lost. This article offers a lot of insightful remarks: [https://learn.microsoft.com/en-us/ef/core/miscellaneous/connection-resiliency](https://learn.microsoft.com/en-us/ef/core/miscellaneous/connection-resiliency)

The option I would go for is not to have retries. Let the EF operation fail, and the user (or API caller) should retry the entire request.

**Comment by batesm** [comment]
Having thought about it I realised that the domain events would have been translated into Outbox message entities and attached to the DbContext. The domain events would not be altogether lost but take a different form. The outbox messages would also be retried. 

It's just something I've learned to be aware of (side effects during a call to SaveChanges(Async) when that call might fail).

As you say if you just retry the entire request (fresh new DbContext each time) then it won't be a problem at all. 

Thanks for the link. I'll have a read.
